File: 9-tree/invert-binary-tree.py
"""
Leet Code 226. Invert Binary Tree
Techniques: Recursion and Iteration (BFS)
Time complexity: O(n)
Space complexity: O(n)
"""

from collections import deque


class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):
    # Recursion
    def invertTree(self, root):
        if not root:
            return root
        root.left, root.right = root.right, root.left
        self.invertTree(root.right)
        self.invertTree(root.left)
        return root

    # Iteration (BFS)
    def invertTree(self, root):
        if not root:
            return
        q = deque()
        q.append(root)
        while q:
            node = q.popleft()
            node.left, node.right = node.right, node.left
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        return root

File: 9-tree/invert-binary-tree.js
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function(root) {
    if (!root) return root;
    let temp = root.left
    root.left = root.right
    root.right = temp
    invertTree(root.left)
    invertTree(root.right)
    return root
};

var invertTree = function(root) {
    if (!root) return root;
    let q = [root]
    while (q.length) {
        let node = q.shift()
        let temp = node.right
        node.right = node.left
        node.left = temp
        if (node.right) q.push(node.right)
        if (node.left) q.push(node.left)
    }
    return root;
};